aten/src/ATen/core/ivalue_inl.h

```diff
@@ -414,7 +414,15 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
     addCallback(
         [fut, cb = std::move(callback)]() {
           try {
-            fut->markCompleted(cb());
+            IValue res = cb();
+            if (res.isFuture()) {
```

Comment:
>  In fact, this suggests that perhaps we should not combine these two separate functions in a single "overloaded" method, but maybe leave them as separate methods (e.g., then and thenAsync? or thenFut?).

agree on the point that we should clearly distinguish these behaviors instead of packing them into the same API. We actually already have a similar thing called `@async_execution` in RPC. I wonder if we should reuse the same decorator?

https://github.com/pytorch/pytorch/blob/dee82ef3ea1ed2cae4e891f071adff03cb4a6e0d/torch/distributed/rpc/functions.py#L4